pkg/services/system_intake_action.go

```diff
@@ -0,0 +1,148 @@
+	submit func(context.Context, *models.SystemIntake) error,
+) func(context.Context, uuid.UUID, *models.ActionType) error {
+	return func(ctx context.Context, intakeID uuid.UUID, actionType *models.ActionType) error {
+		intake, fetchErr := fetch(ctx, intakeID)
+		if fetchErr != nil {
+			return &apperrors.QueryError{
+				Err:       fetchErr,
+				Operation: apperrors.QueryFetch,
+				Model:     intake,
+			}
+		}
+
+		switch *actionType {
+		case models.ActionTypeSUBMIT:
+			return submit(ctx, intake)
+		default:
+			return &apperrors.ResourceConflictError{
+				Err:        errors.New("invalid system intake action type"),
+				Resource:   intake,
+				ResourceID: intake.ID.String(),
+			}
+		}
+	}
+}
+
+// NewSubmitSystemIntake returns a function that
+// executes submit of a system intake
+func NewSubmitSystemIntake(
+	config Config,
+	authorize func(context.Context, *models.SystemIntake) (bool, error),
+	update func(context.Context, *models.SystemIntake) (*models.SystemIntake, error),
+	validateAndSubmit func(context.Context, *models.SystemIntake) (string, error),
+	createAction func(context.Context, *models.Action) (*models.Action, error),
+	fetchUserInfo func(*zap.Logger, string) (*models2.Person, error),
+	emailReviewer func(requester string, intakeID uuid.UUID) error,
+) func(context.Context, *models.SystemIntake) error {
+	return func(ctx context.Context, intake *models.SystemIntake) error {
+		ok, err := authorize(ctx, intake)
+		if err != nil {
+			return err
+		}
+		if !ok {
+			return &apperrors.UnauthorizedError{Err: err}
+		}
+
+		if intake.Status != models.SystemIntakeStatusDRAFT {
+			err := &apperrors.ResourceConflictError{
+				Err:        errors.New("intake is not in DRAFT state"),
+				ResourceID: intake.ID.String(),
+				Resource:   intake,
+			}
+			return err
+		}
+
+		updatedTime := config.clock.Now()
+		intake.UpdatedAt = &updatedTime
+		intake.Status = models.SystemIntakeStatusSUBMITTED
+
+		if intake.AlfabetID.Valid {
+			err := &apperrors.ResourceConflictError{
+				Err:        errors.New("intake has already been submitted to CEDAR"),
+				ResourceID: intake.ID.String(),
+				Resource:   intake,
+			}
+			return err
+		}
+
+		userInfo, err := fetchUserInfo(appcontext.ZLogger(ctx), appcontext.Principal(ctx).ID())
+		if err != nil {
+			return err
+		}
+		if userInfo == nil || userInfo.Email == "" || userInfo.CommonName == "" || userInfo.UserName == "" {
+			return &apperrors.ExternalAPIError{
+				Err:       errors.New("user info fetch was not successful"),
+				Model:     intake,
+				ModelID:   intake.ID.String(),
+				Operation: apperrors.Fetch,
+				Source:    "CEDAR LDAP",
+			}
+		}
+
+		action := models.Action{
+			IntakeID:       intake.ID,
+			ActionType:     models.ActionTypeSUBMIT,
+			ActorName:      userInfo.CommonName,
+			ActorEmail:     userInfo.Email,
+			ActorEUAUserID: userInfo.UserName,
+		}
+		_, err = createAction(ctx, &action)
```

Comment:
The current order of steps here is 
1. Save Action object
2. Submit to CEDAR
3. Save Intake
4. Send email

Is this the best order? considering that any of these failing wouldn't undo the previous ones.

Reply:
I would consider putting CEDAR as the first thing.
The following ordering sort of depends on if you will let a Submit happen if there's already been a Submit processed.